### ISSUE TITLE
feedback for the `strict_top_level_inference` lint

### ISSUE THREAD

**Issue by devoncarew** [dart-lang/sdk#59746]
This is general feedback for the `strict_top_level_inference` lint; cc @srawlins.

In order to try this out, I made a custom version of package:lints/core.yaml and ran it on the sdk repo. Some notes:

- it found 24,835 issues
- 24,374 of them were in test code (so, all, but 100 issues)
- from spot checking them, many of them involved a test method that did have a type annotation (so, tests not as closure but as individual methods; `test_functionTypedFormalParameter() async { ... }`)
- for these cases it would be useful to have a quick fix that added a return type (for many of these cases, this would be `void`, or `Future`, or `Future<void>`).

I think all the cases it fired on were legit. My main feedback from above is that a quick fix would be useful :)

**Comment by srawlins** [comment]
Ah yeah, I noted in the CL that we cannot predict what the missing type _should_ be. But I was only thinking of field types and parameter types.

For return types, we should definitely be able to recommend a type. Maybe not in all cases, but the `void` and `Future<void>` ones should be very easy (no return statements? `void` or `Future<void>`!).

**Comment by devoncarew** [comment]
Yeah, I don't think offering to quick-fix the return type to `dynamic` would be helpful, but if we can determine the right type (`void`, ...), that would be great.

Perhaps the impl. of the quick fix for `always_declare_return_types` would be helpful here?

**Comment by srawlins** [comment]
I forgot to link the CR,  but https://dart-review.googlesource.com/c/sdk/+/401800 links the quick fix that you suggest.

I don't think that this quick fix ever suggests a return type when the function body has no return statements (it would be either `void` or `Future<void>`). I can add that too.

**Comment by frank-weindel** [comment]
My team is looking particularly for a way to prevent any and all inferred `dynamic` typed variables. This linter rule seems to help in many cases for that. But not in others. For example, if you declare a function within another function, `dynamic` is used and allowed if you don't explicitly type a parameter on that function.

**Comment by FMorschel** [comment]
> I don't think that this quick fix ever suggests a return type when the function body has no return statements (it would be either `void` or `Future<void>`). I can add that too.

Yes, it does do this correctly today. No additional work around this is needed. 

> Yeah, I don't think offering to quick-fix the return type to `dynamic` would be helpful, but if we can determine the right type (`void`, ...), that would be great.

When it can't infer a better type (like when you return, for example, `int?` and `String`), it chooses `Object?`. It only ever infers `dynamic` if you are indeed returning a `dynamic` type from what I remember.

P.S.: I'm not sure how to answer the comment above. I just had missed the one I answered just now.

**Comment by srawlins** [comment]
@frank-weindel can you open an issue for what you describe with a minimal reproduction case? Thanks!
